man/diod.8.in
.TH diod 8 "@PACKAGE_VERSION@" "@PACKAGE_NAME@" "@PACKAGE_NAME@"
.SH NAME
diod \- distributed I/O daemon
.SH SYNOPSIS
.B diod
.I "[OPTIONS]"
.SH DESCRIPTION
.B diod
is a 9P2000.L file server.
.LP
Configuration is read from the diod.conf (5) config file.
Some configuration can be overridden on the command line, as described below.
.SH OPTIONS
.TP
.I "-r, --rfdno INT"
.TP
.I "-w, --wfdno INT"
Serve a connected client inherited on the specified read and write
file descriptors.  For file descriptors connected bidirectionally,
the two options may be set to the same value.
.TP
.I "-l, --listen address"
Set the listen address.  The address may be in the form of
HOST:PORT, IP:PORT, or /path/to/unix_domain_socket form (default 0.0.0.0:564).
IPv6 addresses must be enclosed in square brackets.
This option may be specified more than once.
It overrides the \fIlisten\fR config file setting.
.TP
.I "-t, --nwthreads INT"
Set the number of worker threads to spawn to handle 9P operations
for each unique aname.
This option overrides the \fInwthreads\fR setting in diod.conf (5).
The default is 16.
.TP
.I "-e, --export PATH"
Set the file system to be exported.
This option may be specified more than once.
It overrides the \fIexports\fR setting in diod.conf (5).
.TP
.I "-E, --export-all"
Export everything in the server's /proc/mounts.
.TP
.I "-o, --export-opts OPT[,OPT,...]"
Set global export options.
This option overrides the \fIexportopts\fR setting in diod.conf (5).
.TP
.I "-n, --no-auth"
This option allows users to attach without security credentials.
It overrides  the \fIauth_required\fR setting in diod.conf (5).
See SECURITY below.
.TP
.I "-H, --no-hostname-lookup"
This option disables hostname lookups.
It overrides the \fIhostname_lookup\fR setting in diod.conf (5).
.TP
.I "-N, --no-userdb"
This option disables password/group lookups.
It allows any uid to attach and assumes gid=uid, and supplementary groups
contain only the primary gid.
It overrides  the \fIuserdb\fR setting in diod.conf (5).
.TP
.I "-S, --allsquash"
Remap all users to the squash user.
The attaching user need not be present in the server's password file.
This option overrides the \fIallsquash\fR setting in diod.conf (5).
.TP
.I "-U, --squashuser"
Set the squash user.  The default is nobody.
This option overrides the \fIsquashuser\fR setting in diod.conf (5).
Note: if \fBdiod\fR is not run as root, the effective uid overrides
the \fIsquashuser\fR.
.TP
.I "-u, --runas-uid UID"
Run the server as UID and only allow that user to attach.
If invoked as root, \fBdiod\fR sets real and effective uid, gid, and
supplementary groups to those belonging to UID.
.TP
.I "-L, --logdest DEST"
Set the destination for logging.  Possible destinations are
\fIstderr\fR,
\fIstdout\fR, or
a file name.
This option overrides the \fIlogdest\fR setting in diod.conf (5).
.TP
.I "-d, --debug MASK"
Set the debug mask.  The bit values are:
.br
0x01 - log decoded 9P protocol messages
.br
.TP
.I "-c, --config-file PATH"
Set config file path.
.SH SECURITY
\fBdiod\fR optionally uses MUNGE for authentication.
Briefly, a MUNGE credential is a user's uid and gid plus optional payload,
encrypted using a secret shared between client and server, then base64
encoded.
.LP
The 9P2000.L authentication sequence with MUNGE looks like this:
.sp 1
.nf
Tauth afid uname aname n_uname
Rauth aqid
Twrite afid offset count <munge cred>
Rwrite count
Tattach fid afid uname aname n_uname
Rattach qid
Tclunk afid
Rclunk
.fi
.LP
\fITauth\fR and \fITwrite\fR authenticate \fIuname\fR to the server on
\fIafid\fR, which, if the server accepts the credential, can then be used
in the \fITattach\fR to obtain access to the root of a file system represented
by \fIaname\fR as the \fIuname\fR user.  The server knows that all accesses
to fids walked from the attachment fid are being performed by \fIuname\fR.
.LP
If authentication is disabled in the server, the \fITauth\fR may be skipped
and an \fIafid\fR of -1 may be presented in the \fITattach\fR.
.LP
It should be noted that even when authentication is enabled, network
connections between client and server are not protected from eavesdropping
or other attacks.  It is best to use \fBdiod\fR only on networks that are
physically secure.
.LP
When the server is running as an unprivileged user, including when it starts
as root but drops that capability because it is squashing all requests down
to one user, all its file operations are performed as that user.  This is
the simplest and safest mode for \fBdiod\fR to operate in from a security
standpoint.
.LP
Multi-user support is intended to be paired with the Linux v9fs client
and makes assumptions in the interest of correct functioning with v9fs
that may be surprising, notably:
.IP 1.
After \fITattach\fR is accepted for \fIuname=root\fR on a given connection,
subsequent \fITattach\fR requests as other users on the same connection will
be accepted without authentication (\fIafid\fR set to -1).
.IP 2.
Server worker threads handling a request on behalf of a user call setfsuid (2)
and setfsgid (2) to switch to the user's credentials.  Supplementary groups
are also loaded, but only if the server determines that it can do so
in a thread-safe manner.  A warning is issued at server startup if it cannot.
.IP 3.
Server worker threads handling a request on behalf of a non-root user on
a connection authenticated as root set \fBCAP_DAC_OVERRIDE\fR, \fBCAP_CHOWN\fR,
and \fBCAP_FOWNER\fR.  The v9fs client is assumed to be performing access
checks.  This largely works around complications that arise when supplementary
groups cannot be loaded.
.LP
Since the v9fs client does not know how to authenticate with MUNGE,
mount.diod (8) establishes the server connection and performs the
authentication exchange and an initial \fITattach\fR as root, then passes
the open file descriptor into the kernel with the mount (2) system call.
The kernel v9fs client then re-introduces itself as root with \fIafid\fR
set to -1.  As noted above this \fITattach\fR and subsequent ones on this
connection are accepted by the server by virtue of the initial authenticated
root attachment.
.SH "FILES"
@X_SBINDIR@/diod
.br
@X_SYSCONFDIR@/diod.conf
.SH "SEE ALSO"
diod.conf (5), mount.diod (8), hosts_access (5)


NEWS
=========================================================================
Release Notes for DIOD version 1.0.24                         31 Mar 2015
=========================================================================

* Replace sysv-init script with systemd unit file

* Move fully to github (including docs)

* Misc. minor updates, mostly tests and build environment, for rhel7

=========================================================================
Release Notes for DIOD version 1.0.23                         26 Aug 2014
=========================================================================

* Add dist tag to release (TOSS-2662)

* Minor autotools forward-portability (issue 124)

* Doc: AUTHORS, README, license headers updated

* Stop tracking additional autotools build products.

=========================================================================
Release Notes for DIOD version 1.0.22                         19 Aug 2014
=========================================================================

* Stop tracking autotools build products in git.

* Fix assertion failure at srv.c:754 (issue 122)

* Portability fixes for musl libc (Felix Janda)

* Change default rdma port to 5640 to match kernel

* rdmatrans: only register the used buffer to the network
  adapter (Dominique Martinet)

* fix configure --with-tcmalloc logic

* libnpfs: fids cloned in xattrwalk should inherit flags

* rdmatrans: compilation error

=========================================================================
Release Notes for DIOD version 1.0.21                         31 May 2013
=========================================================================

* Require that "ctl" be explicitly exported.  This allows export options
  such as "ro", "noauth", and "privport" to apply to the ctl synthetic
  file system.

* The "auth_required_ctl" config option introduced in 1.0.15 is now
  deprecated.

* Add test coverage for trusted.* and security.* xattr namespaces.

* Add more mmap test coverage.

=========================================================================
Release Notes for DIOD version 1.0.20                         17 May 2013
=========================================================================

* Fix diodshowmount segfault on exit with 'noauth'.

* Add diodls --privport option.

* Additional test coverage (atomic create test, pathwalk test)

=========================================================================
Release Notes for DIOD version 1.0.19                         07 May 2013
=========================================================================

* Additional test coverage for xattrs:
  - Verify that setxattr flags XATTR_CREATE, XATTR_REPLACE work.
  - Verify that that removal of a nonexistent user xattr returns ENODATA

* Fix failure of tests/user/t17 when SELinux is enabled (bug in the test)

=========================================================================
Release Notes for DIOD version 1.0.18                         06 May 2013
=========================================================================

* Man page update.

=========================================================================
Release Notes for DIOD version 1.0.17                         06 May 2013
=========================================================================

* RPM: add missing BuildRequires for 'libattr-devel'.

=========================================================================
Release Notes for DIOD version 1.0.15                         03 May 2013
=========================================================================

* Add xattr support (issue 95)

* Support for dracut bootstrap of 9nbd network block device:
  - New 'dioddate' utility to synchronize system time with server's,
    and 'auth_required_ctl' config option which allows MUNGE
    authentication to be disabled for the ctl synthetic file system.
  - Add 'noauth' and 'privport' export options which disable MUNGE auth,
    and require attaches to originate from a privileged port (512-1024)
    for a specific export.    

=========================================================================
Release Notes for DIOD version 1.0.14                         12 Jul 2012
=========================================================================

* Add mount.diod --attach and --detach options to support 9nbd
  (9P network block device).

* Allow non-directories to be exported (issue 113)

* Return Rlerror(ENOENT) in response to Tauth if auth not required
  instead of Rlerror(0), which confused kernel 9P code.
    
=========================================================================
Release Notes for DIOD version 1.0.13                         22 May 2012
=========================================================================

* Allow DIOD_MOUNTOPTS in /etc/sysconfig/auto.diod

  Normalize -s,--server NAME option in diodcat, diodload, diodls,
  and diodshowmount utilities, where NAME can be IP[:PORT], HOST[:PORT],
  or /path/to/socket.

* Add diodshowmount utility (issue 111)

diod/diod.c
/************************************************************\
 * Copyright 2010 Lawrence Livermore National Security, LLC
 * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
 *
 * This file is part of the diod 9P server project.
 * For details, see https://github.com/chaos/diod.
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
\************************************************************/

/* diod.c - distributed I/O daemon */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <poll.h>
#ifndef _BSD_SOURCE
#define _BSD_SOURCE         /* daemon () */
#endif
#include <stdlib.h>
#include <stdint.h>
#include <sys/types.h>
#include <unistd.h>
#if USE_IMPERSONATION_LINUX
#include <sys/syscall.h>
#endif
#include <stdio.h>
#if HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <errno.h>
#include <pwd.h>
#include <grp.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/resource.h>
#if HAVE_SYS_PRCTL_H
#include <sys/prctl.h>
#endif
#include <string.h>
#include <signal.h>
#include <pthread.h>

#include "9p.h"
#include "npfs.h"
#include "list.h"

#include "diod_log.h"
#include "diod_conf.h"
#include "diod_sock.h"
#if WITH_RDMATRANS
#include "diod_rdma.h"
#endif

#include "ops.h"

#if USE_IMPERSONATION_GANESHA
#include "ganesha-syscalls.h"
#endif

typedef enum { SRV_FILEDES, SRV_SOCKTEST, SRV_NORMAL } srvmode_t;

static void          _daemonize (void);
static void          _setrlimit (void);
static void          _become_user (char *name, uid_t uid, int realtoo);
static void          _service_run (srvmode_t mode, int rfdno, int wfdno);

#ifndef NR_OPEN
#define NR_OPEN         1048576 /* works on RHEL 5 x86_64 arch */
#endif

#define OPTIONS "fr:w:d:l:t:e:Eo:u:SL:nHpc:NU:s"

#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long (ac,av,opt,lopt,NULL)
static const struct option longopts[] = {
    {"foreground",         no_argument,        0, 'f'},
    {"rfdno",              required_argument,  0, 'r'},
    {"wfdno",              required_argument,  0, 'w'},
    {"debug",              required_argument,  0, 'd'},
    {"listen",             required_argument,  0, 'l'},
    {"nwthreads",          required_argument,  0, 't'},
    {"export",             required_argument,  0, 'e'},
    {"export-all",         no_argument,        0, 'E'},
    {"export-opts",        required_argument,  0, 'o'},
    {"no-auth",            no_argument,        0, 'n'},
    {"no-hostname-lookup", no_argument,        0, 'H'},
    {"statfs-passthru",    no_argument,        0, 'p'},
    {"no-userdb",          no_argument,        0, 'N'},
    {"runas-uid",          required_argument,  0, 'u'},
    {"allsquash",          no_argument,        0, 'S'},
    {"squashuser",         required_argument,  0, 'U'},
    {"logdest",            required_argument,  0, 'L'},
    {"config-file",        required_argument,  0, 'c'},
    {"socktest",           no_argument,        0, 's'},
    {0, 0, 0, 0},
};
#else
#define GETOPT(ac,av,opt,lopt) getopt (ac,av,opt)
#endif

static void
usage()
{
    fprintf (stderr,
"Usage: diod [OPTIONS]\n"
"   -f,--foreground         do not fork and disassociate with tty\n"
"   -r,--rfdno              service connected client on read file descriptor\n"
"   -w,--wfdno              service connected client on write file descriptor\n"
"   -l,--listen IP:PORT     set interface to listen on (multiple -l allowed)\n"
"   -t,--nwthreads INT      set number of I/O worker threads to spawn\n"
"   -e,--export PATH        export PATH (multiple -e allowed)\n"
"   -E,--export-all         export all mounted file systems\n"
"   -o,--export-opts        set global export options (comma-seperated)\n"
"   -n,--no-auth            disable authentication check\n"
"   -H,--no-hostname-lookup disable hostname lookups\n"
"   -p,--statfs-passthru    statfs should return underly f_type not V9FS_MAGIC\n"
"   -N,--no-userdb          bypass password/group file lookup\n"
"   -u,--runas-uid UID      only allow UID to attach\n"
"   -S,--allsquash          map all users to the squash user\n"
"   -U,--squashuser USER    set the squash user (default nobody)\n"
"   -L,--logdest DEST       log to DEST, can be syslog, stderr, or file\n"
"   -d,--debug MASK         set debugging mask\n"
"   -c,--config-file FILE   set config file path\n"
"   -s,--socktest           run in test mode where server exits early\n"
    );
    exit (1);
}

int
main(int argc, char **argv)
{
    int c;
    char *copt = NULL;
    srvmode_t mode = SRV_NORMAL;
    int rfdno = -1, wfdno = -1;

    diod_log_init (argv[0]);
    diod_conf_init ();

    /* config file overrides defaults */
    opterr = 0;
    while ((c = GETOPT (argc, argv, OPTIONS, longopts)) != -1) {
        switch (c) {
            case 'c':   /* --config-file PATH */
                copt = optarg;
                break;
            default:
                break;
        }
    }
    diod_conf_init_config_file (copt);

    /* Command line overrides config file.
     */
    optind = 0;
    opterr = 0;
    while ((c = GETOPT (argc, argv, OPTIONS, longopts)) != -1) {
        switch (c) {
            case 'f':   /* --foreground */
                diod_conf_set_foreground (1);
                break;
            case 'r':   /* --rfdno */
                mode = SRV_FILEDES;
                rfdno = strtoul (optarg, NULL, 10);
                break;
            case 'w':   /* --wfdno */
                mode = SRV_FILEDES;
                wfdno = strtoul (optarg, NULL, 10);
                break;
            case 'd':   /* --debug MASK */
                diod_conf_set_debuglevel (strtoul (optarg, NULL, 0));
                break;
            case 'l':   /* --listen HOST:PORT or /path/to/socket */
                if (!diod_conf_opt_listen ())
                    diod_conf_clr_listen ();
                if (!strchr (optarg, ':') && optarg[0] != '/')
                    usage ();
                if (optarg[0] == '[') {
                    char *end = strchr (optarg, ']');

                    if (!end || !strchr (end, ':'))
                        usage ();
                }
                diod_conf_add_listen (optarg);
                break;
            case 't':   /* --nwthreads INT */
                diod_conf_set_nwthreads (strtoul (optarg, NULL, 10));
                break;
            case 'c':   /* --config-file PATH */
                break;
            case 'e':   /* --export PATH */
                if (!diod_conf_opt_exports ())
                    diod_conf_clr_exports ();
                diod_conf_add_exports (optarg);
                break;
            case 'E':   /* --export-all */
                diod_conf_set_exportall (1);
                break;
            case 'o':   /* --export-ops opt,[opt,...] */
                diod_conf_set_exportopts (optarg);
                break;
            case 'n':   /* --no-auth */
                diod_conf_set_auth_required (0);
                break;
            case 'H':   /* --no-hostname-lookup */
                diod_conf_set_hostname_lookup (0);
                break;
            case 'p':   /* --statfs-passthru */
                diod_conf_set_statfs_passthru (0);
                break;
            case 'N':   /* --no-userdb */
                diod_conf_set_userdb (0);
                break;
            case 'S':   /* --allsquash */
                diod_conf_set_allsquash (1);
                break;
            case 'U':   /* --squashuser USER */
                diod_conf_set_squashuser (optarg);
                break;
            case 'u': { /* --runas-uid UID */
                uid_t uid;
                char *end;

                errno = 0;
                uid = strtoul (optarg, &end, 10);
                if (errno != 0)
                    err_exit ("error parsing --runas-uid argument");
                if (*end != '\0')
                    msg_exit ("error parsing --runas-uid argument");
                diod_conf_set_runasuid (uid);
                break;
            }
            case 's':   /* --socktest */
                mode = SRV_SOCKTEST;
                break;
            case 'L':   /* --logdest DEST */
                diod_conf_set_logdest (optarg);
                diod_log_set_dest (optarg);
                break;
            default:
                usage();
        }
    }
    if (optind < argc)
        usage();
    if (diod_conf_opt_runasuid () && diod_conf_get_allsquash ())
        msg_exit ("--runas-uid and allsquash cannot be used together");
    if (mode == SRV_FILEDES && (rfdno == -1 || wfdno == -1))
        msg_exit ("--rfdno,wfdno must be used together");

    diod_conf_validate_exports ();

    if (geteuid () == 0)
        _setrlimit ();

    _service_run (mode, rfdno, wfdno);

    diod_conf_fini ();
    diod_log_fini ();

    exit (0);
}

/* Switch to user/group, load the user's supplementary groups.
 * Print message and exit on failure.
 */
static void
_become_user (char *name, uid_t uid, int realtoo)
{
    int err;
    struct passwd pw, *pwd;
    int len = sysconf(_SC_GETPW_R_SIZE_MAX);
    char *buf;
    int nsg;
    gid_t sg[64];
    char *endptr;

    if (len == -1)
        len = 4096;
    if (!(buf = malloc(len)))
        msg_exit ("out of memory");
    if (name) {
        errno = 0;
        uid = strtoul (name, &endptr, 10);
        if (errno == 0 && *name != '\0' && *endptr == '\0')
            name = NULL;
    }
    if (name) {
        if ((err = getpwnam_r (name, &pw, buf, len, &pwd)) != 0)
            errn_exit (err, "error looking up user %s", name);
        if (!pwd)
            msg_exit ("error looking up user %s", name);
    } else {
        if ((err = getpwuid_r (uid, &pw, buf, len, &pwd)) != 0)
            errn_exit (err, "error looking up uid %d", uid);
        if (!pwd)
            msg_exit ("error looking up uid %d", uid);
    }
    nsg = sizeof (sg) / sizeof(sg[0]);
    if (getgrouplist(pwd->pw_name, pwd->pw_gid, sg, &nsg) == -1)
        err_exit ("user is in too many groups");
#if USE_IMPERSONATION_LINUX
    if (syscall(SYS_setgroups, nsg, sg) < 0)
        err_exit ("setgroups");
#else
    if (setgroups (nsg, sg) < 0)
        err_exit ("setgroups");
#endif
    if (setregid (realtoo ? pwd->pw_gid : -1, pwd->pw_gid) < 0)
        err_exit ("setreuid");
    if (setreuid (realtoo ? pwd->pw_uid : -1, pwd->pw_uid) < 0)
        err_exit ("setreuid");
    free (buf);
}

/* Remove any resource limits.
 * Exit on error
 */
static void
_setrlimit (void)
{
    struct rlimit r, r2;

    r.rlim_cur = r.rlim_max = RLIM_INFINITY;
    if (setrlimit (RLIMIT_FSIZE, &r) < 0)
        err_exit ("setrlimit RLIMIT_FSIZE");

    r.rlim_cur = r.rlim_max = NR_OPEN;
    r2.rlim_cur = r2.rlim_max = sysconf(_SC_OPEN_MAX);
    if (setrlimit (RLIMIT_NOFILE, &r) < 0)
        if (errno != EPERM || setrlimit (RLIMIT_NOFILE, &r2) < 0)
            err_exit ("setrlimit RLIMIT_NOFILE");

    r.rlim_cur = r.rlim_max = RLIM_INFINITY;
#ifdef RLIMIT_LOCKS
    if (setrlimit (RLIMIT_LOCKS, &r) < 0)
        err_exit ("setrlimit RLIMIT_LOCKS");
#endif
    r.rlim_cur = r.rlim_max = RLIM_INFINITY;
    if (setrlimit (RLIMIT_CORE, &r) < 0)
        err_exit ("setrlimit RLIMIT_CORE");

    r.rlim_cur = r.rlim_max = RLIM_INFINITY;
    if (setrlimit (RLIMIT_AS, &r) < 0)
        err_exit ("setrlimit RLIMIT_AS");

}

/* Create run directory if it doesn't exist and chdir there.
 * Disassociate from parent's controlling tty.  Switch logging to syslog.
 * Exit on error.
 */
static void
_daemonize (void)
{
    char rdir[PATH_MAX];

    snprintf (rdir, sizeof(rdir), "%s/run/diod", X_LOCALSTATEDIR);
    if (mkdir (rdir, 0755) < 0 && errno != EEXIST) {
        msg ("failed to find/create %s, running out of /tmp", rdir);
        snprintf (rdir, sizeof(rdir), "/tmp");
    }
    if (chdir (rdir) < 0)
        err_exit ("chdir %s", rdir);
    if (daemon (1, 0) < 0)
        err_exit ("daemon");
}

/**
 ** Service startup
 **/

struct svc_struct {
    srvmode_t mode;
    int rfdno;
    int wfdno;
    Npsrv *srv;
    struct pollfd *fds;
    int nfds;
    pthread_t t;
    int shutdown;
    int reload;
#if WITH_RDMATRANS
    diod_rdma_t rdma;
    pthread_t rdma_t;
#endif
};
static struct svc_struct ss;

static void
_sighand (int sig)
{
    switch (sig) {
        case SIGHUP:
            msg ("caught SIGHUP: reloading config");
            ss.reload = 1;
            break;
        case SIGTERM:
            msg ("caught SIGTERM: shutting down");
            ss.shutdown = 1;
            break;
        case SIGUSR1:
            ss.shutdown = 1;
            break;
        case SIGUSR2:
            break;
        default:
            msg ("caught signal %d: ignoring", sig);
            break;
    }
}


/* Thread to handle SIGHUP, SIGTERM, and new connections on listen ports.
 */
static void *
_service_loop (void *arg)
{
    int lookup = diod_conf_get_hostname_lookup ();
    sigset_t sigs;
    int i;

    sigfillset (&sigs);
    sigdelset (&sigs, SIGHUP);
    sigdelset (&sigs, SIGTERM);
    sigdelset (&sigs, SIGUSR1);

    switch (ss.mode) {
        case SRV_FILEDES:
            diod_sock_startfd (ss.srv, ss.rfdno, ss.wfdno, "stdin", 0);
            break;
        case SRV_NORMAL:
        case SRV_SOCKTEST:
            break;
    }
    while (!ss.shutdown) {
        if (ss.reload) {
            diod_conf_init_config_file (NULL);
            np_usercache_flush (ss.srv);
            ss.reload = 0;
        }
        for (i = 0; i < ss.nfds; i++) {
            ss.fds[i].events = POLLIN;
            ss.fds[i].revents = 0;
        }
        if (ppoll (ss.fds, ss.nfds, NULL, &sigs) < 0) {
            if (errno == EINTR)
                continue;
            err_exit ("ppoll");
        }
        for (i = 0; i < ss.nfds; i++) {
            if ((ss.fds[i].revents & POLLIN)) {
                diod_sock_accept_one (ss.srv, ss.fds[i].fd, lookup);
            }
        }
    }
    return NULL;
}

#if WITH_RDMATRANS
static void *
_service_loop_rdma (void *arg)
{
    while (!ss.shutdown) {
        msg ("waiting on rdma connection");
        diod_rdma_accept_one (ss.srv, ss.rdma);
    }
    return NULL;
}
#endif

/* Set up signal handlers for SIGHUP and SIGTERM and block them.
 * Threads will inherit this signal mask; _service_loop () will unblock.
 * Install handler for SIGUSR2 and don't block - this signal is used to
 * interrupt I/O operations when handling a 9p flush.
 */
static void
_service_sigsetup (void)
{
    struct sigaction sa;
    sigset_t sigs;

    sa.sa_flags = 0;
    sigemptyset (&sa.sa_mask);
    sa.sa_handler = _sighand;
    if (sigaction (SIGHUP, &sa, NULL) < 0)
        err_exit ("sigaction");
    if (sigaction (SIGTERM, &sa, NULL) < 0)
        err_exit ("sigaction");
    if (sigaction (SIGUSR1, &sa, NULL) < 0)
        err_exit ("sigaction");
    if (sigaction (SIGUSR2, &sa, NULL) < 0)
        err_exit ("sigaction");

    sigemptyset (&sigs);
    sigaddset (&sigs, SIGHUP);
    sigaddset (&sigs, SIGTERM);
    sigaddset (&sigs, SIGUSR1);
    if ((sigprocmask (SIG_BLOCK, &sigs, NULL) < 0))
        err_exit ("sigprocmask");
}

#if USE_IMPERSONATION_LINUX
/* Test whether setgroups applies to thread or process.
 * On RHEL6 (2.6.32 based) it applies to threads and we can use it.
 * On Ubuntu 11 (2.6.38 based) it applies to the whole process and we can't.
 */
void *
_test_setgroups_thread (void *arg)
{
    gid_t sg[] = { 42, 37, 63 };

    if (syscall(SYS_setgroups, 3, sg) < 0)
        err_exit ("setgroups");
    return NULL;
}
static int
_test_setgroups (void)
{
    long ngroups_max = sysconf(_SC_NGROUPS_MAX);
    gid_t *sg;
    int nsg, n;
    int rc = 0;
    pthread_t t;
    int err;

    if (ngroups_max < 0)
        ngroups_max = 64;
    if (!(sg = malloc (ngroups_max * sizeof (gid_t))))
        msg_exit ("out of memory");
    if ((nsg = getgroups (ngroups_max, sg)) < 0)
        err_exit ("getgroups");
    if (syscall(SYS_setgroups, 0, NULL) < 0) /* clear groups */
        err_exit ("setgroups");
    if ((err = pthread_create (&t, NULL, _test_setgroups_thread, NULL)))
        err_exit ("pthread_create");
    if ((err = pthread_join (t, NULL)))
        err_exit ("pthread_join");
    if ((n = getgroups (ngroups_max, sg)) < 0)
        err_exit ("getgroups");
    if (n == 0)
        rc = 1;
    if (syscall(SYS_setgroups, nsg, sg) < 0)
        err_exit ("setgroups");
    free (sg);
    return rc;
}
#endif /* USE_IMPERSONATION_LINUX */

static void
_service_run (srvmode_t mode, int rfdno, int wfdno)
{
    List l = diod_conf_get_listen ();
    int nwthreads = diod_conf_get_nwthreads ();
    int flags = diod_conf_get_debuglevel ();
    uid_t euid = geteuid ();
    int n;

    ss.mode = mode;
    ss.rfdno = rfdno;
    ss.wfdno = wfdno;
    ss.shutdown = 0;
    ss.reload = 0;
    _service_sigsetup ();

    ss.fds = NULL;
    ss.nfds = 0;
    switch (mode) {
        case SRV_FILEDES:
            break;
        case SRV_NORMAL:
        case SRV_SOCKTEST:
            if (!diod_sock_listen (l, &ss.fds, &ss.nfds))
                msg_exit ("failed to set up listener");
#if WITH_RDMATRANS
            ss.rdma = diod_rdma_create ();
            diod_rdma_listen (ss.rdma);
#endif
            break;
    }

    /* manipulate squash/runas users if not root */
    if (euid != 0) {
        if (diod_conf_get_allsquash ()) {
            struct passwd *pw = getpwuid (euid);
            char *su = diod_conf_get_squashuser ();
            if (!pw)
                msg_exit ("getpwuid on effective uid failed");
            if (strcmp (pw->pw_name, su) != 0) {
                if (strcmp (su, DFLT_SQUASHUSER) != 0)
                    msg ("changing squashuser '%s' to '%s' "
                         "since you are not root", su, pw->pw_name);
                diod_conf_set_squashuser (pw->pw_name); /* fixes issue 41 */
            }
        } else { /* N.B. runasuser cannot be set in the config file */
            uid_t ruid = diod_conf_get_runasuid ();
            if (diod_conf_opt_runasuid () && ruid != euid)
                msg ("changing runasuid %d to %d "
                     "since you are not root", ruid, euid);
            diod_conf_set_runasuid (euid);
        }
    }

    if (!diod_conf_get_foreground () && mode != SRV_FILEDES)
        _daemonize (); /* implicit fork - no pthreads before this */
    if (!diod_conf_get_foreground () && mode != SRV_FILEDES)
        diod_log_set_dest (diod_conf_get_logdest ());

    /* drop root */
    if (euid == 0) {
        if (diod_conf_get_allsquash ())
            _become_user (diod_conf_get_squashuser (), -1, 1);
        else if (diod_conf_opt_runasuid ())
            _become_user (NULL, diod_conf_get_runasuid (), 1);
    }

    /* clear umask */
    umask (0);

    flags |= SRV_FLAGS_LOOSEFID;        /* work around buggy clients */
    flags |= SRV_FLAGS_AUTHCONN;
    //flags |= SRV_FLAGS_FLUSHSIG;      /* XXX temporarily off */
    if (geteuid () == 0) {
        flags |= SRV_FLAGS_SETFSID;
        flags |= SRV_FLAGS_DAC_BYPASS;
#if USE_IMPERSONATION_LINUX
        if (_test_setgroups ())
            flags |= SRV_FLAGS_SETGROUPS;
        else
            msg ("test_setgroups: groups are per-process (disabling)");
#elif USE_IMPERSONATION_GANESHA
        if (init_ganesha_syscalls() < 0)
            msg ("nfs-ganesha-kmod not loaded: changing user/group will fail");
        /* SRV_FLAGS_SETGROUPS is ignored in user-freebsd.c */
#else
        msg ("warning: cannot change user/group (built with --disable-impersonation)");
#endif
    }

    /* Process dumpable flag may have been cleared by uid manipulation above.
     * Set it here, then maintain it in user.c::np_setfsid () as uids are
     * further manipulated.
     */
#if HAVE_SYS_PRCTL_H
    if (prctl (PR_SET_DUMPABLE, 1, 0, 0, 0) < 0)
        err_exit ("prctl PR_SET_DUMPABLE failed");
#endif

    if (!diod_conf_get_userdb ())
        flags |= SRV_FLAGS_NOUSERDB;
    if (!(ss.srv = np_srv_create (nwthreads, flags))) /* starts threads */
        errn_exit (np_rerror (), "np_srv_create");
    if (diod_init (ss.srv) < 0)
        errn_exit (np_rerror (), "diod_init");

    if ((n = pthread_create (&ss.t, NULL, _service_loop, NULL)))
        errn_exit (n, "pthread_create _service_loop");
#if WITH_RDMATRANS
    if ((n = pthread_create (&ss.rdma_t, NULL, _service_loop_rdma, NULL)))
        errn_exit (n, "pthread_create _service_loop_rdma");
#endif
    switch (mode) {
        case SRV_FILEDES:
        case SRV_SOCKTEST:
            np_srv_wait_conncount (ss.srv, 1);
            pthread_kill (ss.t, SIGUSR1);
            break;
        case SRV_NORMAL:
            break;
    }
    if ((n = pthread_join (ss.t, NULL)))
        errn_exit (n, "pthread_join _service_loop");
#if WITH_RDMATRANS
    if ((n = pthread_join (ss.rdma_t, NULL)))
        errn_exit (n, "pthread_join _service_loop_rdma");
#endif

    np_srv_shutdown(ss.srv);
    diod_fini (ss.srv);
    np_srv_destroy (ss.srv);
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
